### PR TITLE
New rule: cascade_invocations.

### PR DESCRIPTION
**Pull request by alexeieleusis**
Cascade consecutive method invocations on the same reference.

### CODE REVIEW

**Comment by alexeieleusis** [comment]
ptal @bwilkerson @pq

File: lib/src/rules/cascade_invocations.dart [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  bool _previousIsMethodInvocation(Iterable previousChilds) =>
+      previousChilds.isNotEmpty &&
+      previousChilds.last is ExpressionStatement &&
+      previousChilds.last.expression is MethodInvocation;
+
+  bool _previousIsSingleVariableDeclaration(
+          Iterable previousChilds, VariableDeclarationStatement variables) =>
+      previousChilds.isNotEmpty &&
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: I don't understand why we need to test that the collection is not empty. And I don't think that we need to test that `variables` is not null. If we can remove both tests, then this method could be inlined.

File: lib/src/rules/cascade_invocations.dart [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: Consider visiting ExpressionStatement rather than MethodInvocation. That way you won't have to test the parent of every method invocation nested within other expressions.

File: test/rules/cascade_invocations.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+class HasMethodWithNoCascade {
+  List<int> list = [];
+
+  void noCascade() {
+    list.removeWhere((i) => i % 5 == 0);
+    list.clear(); // LINT
+  }
+}
```
Comment (bwilkerson) [inline test/rules/cascade_invocations.dart]: You need to test negative cases, such as:
```
foo.m();
bar.m();
```
or
```
String s = '';
x.m();
```
where it matches the lexical pattern but doesn't refer to the same variable. You should also test that it work correctly when the invocation is nested in a statement or expression:
```
if (x.m() != null) ...
```

File: lib/src/rules/cascade_invocations.dart [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/linter.dart';
+
+const _desc = r'Cascade consecutive method invocations on the same reference.';
+
+const _details = r'''
+
+**DO** Use the cascading style when succesively invoking methods on the same
+ reference.
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: Do we care about getters? For example:
```
x.m();
x.g;
x.m();
```

File: lib/src/rules/cascade_invocations.dart (hunk 2 above)
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: done

File: lib/src/rules/cascade_invocations.dart [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (node.parent is! ExpressionStatement ||
```
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: Refactored and applied where relevant.

File: lib/src/rules/cascade_invocations.dart [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        node.realTarget is! SimpleIdentifier) {
+      return;
+    }
+
+    final simpleIdentifier = node.realTarget as SimpleIdentifier;
+    Block body = node.parent.parent;
+    Iterable previousChilds =
```
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: Fixed.

File: lib/src/rules/cascade_invocations.dart (hunk 1 above)
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: IIRC if the collection is empty and we try to access the last element an exception will be thrown. In any case refactored and inlined as suggested.

File: lib/src/rules/cascade_invocations.dart [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+          Iterable previousChilds, VariableDeclarationStatement variables) =>
+      previousChilds.isNotEmpty &&
+      variables != null &&
+      variables.variables.variables.length == 1;
+
+  void _reportIfConsecutiveInvocationsOnSameTarget(Iterable previousChilds,
+      SimpleIdentifier simpleIdentifier, MethodInvocation node) {
+    ExpressionStatement previousStatement = previousChilds.last;
+    MethodInvocation previousInvocation = previousStatement.expression;
+    if (previousInvocation.realTarget is SimpleIdentifier &&
```
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: Abstracted to variables wherever made sense, hopefully I didn't miss occurrences.

File: test/rules/cascade_invocations.dart (hunk 3 above)
Comment (alexeieleusis) [inline test/rules/cascade_invocations.dart]: Done

**Comment by alexeieleusis** [comment]
This rule is on the fast side and I think requires no resolution so it could be integrated with tools that have that restriction. cc @matanlurey

File: lib/src/rules/cascade_invocations.dart (hunk 1 above)
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: You recall correctly, but in the only place this method was called we had already accessed the last element, so we knew that it couldn't be empty.

File: lib/src/rules/cascade_invocations.dart [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+  @override
+  void visitExpressionStatement(ExpressionStatement node) {
+    final prefixIdentifier = _findTarget(node);
+    if (prefixIdentifier == null) {
+      return;
+    }
+
+    final Block body = node.getAncestor((n) => n is Block);
+    final previousNodes = body.childEntities.takeWhile((n) => n != node);
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: If you make the change above, then you can look at the statements in the block rather than the childEntities.

File: lib/src/rules/cascade_invocations.dart [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  _Visitor(this.rule);
+
+  @override
+  void visitExpressionStatement(ExpressionStatement node) {
+    final prefixIdentifier = _findTarget(node);
+    if (prefixIdentifier == null) {
+      return;
+    }
+
+    final Block body = node.getAncestor((n) => n is Block);
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: You don't actually need / want to do this. If the statement isn't the direct child of the block, then all of the rest of this method can be skipped.

File: lib/src/rules/cascade_invocations.dart [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    final previousNodes = body.childEntities.takeWhile((n) => n != node);
+    if (previousNodes.isEmpty) {
+      return;
+    }
+
+    final previousNode = previousNodes.last;
+    if (previousNode is! AstNode) {
```
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: And if you make the changes above then you don't need to test whether it's an AstNode.

File: lib/src/rules/cascade_invocations.dart (hunk 4 above)
Comment (bwilkerson) [inline lib/src/rules/cascade_invocations.dart]: Perhaps update the documentation to include getters? Also, do we care about setters:
```
x.m();
x.s = '';
x.m();
```

**Comment by bwilkerson** [comment]
> This rule is on the fast side and I think requires no resolution ...

Unfortunately, it does. If you use staticElement, then it requires resolution.

File: lib/src/rules/cascade_invocations.dart (hunk 10 above)
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: :D I really disliked this comparison.

File: lib/src/rules/cascade_invocations.dart (hunk 8 above)
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: Fixed.

File: lib/src/rules/cascade_invocations.dart (hunk 9 above)
Comment (alexeieleusis) [inline lib/src/rules/cascade_invocations.dart]: Changed for parent.

**Comment by alexeieleusis** [comment]
Friendly ping, I think I addressed open issues.

**Comment by bwilkerson** [comment]
lgtm
